src/Workspaces/Remote/Core/Serialization/MessagePackFormatters.cs

```diff
@@ -70,9 +78,26 @@ internal sealed class ProjectIdFormatter : IMessagePackFormatter<ProjectId?>
                         return null;
                     }
 
-                    Contract.ThrowIfFalse(reader.ReadArrayHeader() == 2);
+                    var arrayCount = reader.ReadArrayHeader();
+                    Contract.ThrowIfFalse(arrayCount is 1 or 2);
                     var id = GuidFormatter.Instance.Deserialize(ref reader, options);
-                    var debugName = reader.ReadString();
+                    string? debugName;
+
+                    if (arrayCount == 1)
+                    {
+                        // This ProjectId has previously been deserialized. Attempt to find it's
+                        // debugName from the cache. This *should* always succeed, but if not, it's
+                        // ok to proceed with a null debugName.
+                        _debugNames.TryGetValue(id, out debugName);
```

Comment:
Don't love this. But I'm willing to try it out.  Seems like we could have a small race where some IDs don't have a debug name. But ideally it's a very small amount.